common/unified/base/kernel_launch.hpp

```diff
@@ -184,75 +238,139 @@ struct matrix_accessor {
  *
  * By default, it only maps std::complex to the corresponding device
  * representation of the complex type. There are specializations for dealing
- * with gko::array and gko::matrix::Dense (both const and mutable) that map them
+ * with gko::array and gko::matrix::Dense that map them
  * to plain pointers or matrix_accessor objects.
  *
- * @tparam T  the type being mapped. It will be used based on a
- *            forwarding-reference, i.e. preserve references in the input
- *            parameter, so special care must be taken to only return types that
- *            can be passed to the device, i.e. (structs containing) device
- *            pointers or values. This means that T will be either a r-value or
- *            l-value reference.
+ * @tparam T  the underlying type being mapped. Any references or const
+ *            qualifiers have to be resolved before passing the type.
+ *            The distinction between const/mutable objects is done by
+ *            overloading the map_to_device function.
+ * @tparam PtrWrapper  the pointer type. By default, it's just `T*`, but it may
+ *                     be set to restricted_ptr.
```

Comment:
I've changed a bit how `to_device_type_impl<T>` is implemented, since it was easier for me to reason about it after removing all cv/ref.
If wanted I can also revert this change.